yt_dlp/extractor/telegram.py

```diff
@@ -1,41 +1,141 @@
+        },
+        'playlist_count': 2,
+        'params': {
+            'noplaylist': False,
+            'skip_download': True,
+        }
     }]
 
     def _real_extract(self, url):
-        video_id = self._match_id(url)
-        webpage = self._download_webpage(url, video_id, query={'embed': 0})
-        webpage_embed = self._download_webpage(url, video_id, query={'embed': 1}, note='Downloading ermbed page')
+        uploader, msg_id = self._match_valid_url(url).group('uploader', 'id')
 
-        formats = [{
-            'url': self._proto_relative_url(self._search_regex(
-                '<video[^>]+src="([^"]+)"', webpage_embed, 'source')),
-            'ext': 'mp4',
-        }]
-        self._sort_formats(formats)
-
-        return {
-            'id': video_id,
-            'title': self._html_search_meta(['og:title', 'twitter:title'], webpage, default=None),
-            'description': self._html_search_meta(
-                ['og:description', 'twitter:description'], webpage,
-                default=clean_html(get_element_by_class('tgme_widget_message_text', webpage_embed))),
-            'thumbnail': self._search_regex(
-                r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
-                webpage_embed, 'thumbnail'),
-            'formats': formats,
+        query = {'embed': '1'}
+        if self.get_param('noplaylist'):
+            query['single'] = True
+        elif self.get_param('noplaylist') is False:
+            pass
+        elif '?single' in url:  # param 'noplaylist' is None
+            query['single'] = True
+        url = update_url_query(url, {'single': []})  # strip 'single' from original URL query
+        embed = self._download_webpage(url, msg_id, query=query, note='Downloading embed frame')
+
+        message = {
+            'title': clean_html(get_element_by_class('tgme_widget_message_author', embed)),
+            'description': clean_html(get_element_by_class('tgme_widget_message_text', embed)),
+            'uploader': uploader,
+            'timestamp': unified_timestamp(self._search_regex(
+                r'<time[^>]*datetime="([^"]*)"', embed, 'timestamp', fatal=False)),
         }
+
+        videos = []
+        for video in re.findall(r'<a class="tgme_widget_message_video_player(?s:.+?)</time>', embed):
+            video_url = self._search_regex(
+                r'<video[^>]+src="([^"]+)"', video, 'video URL', fatal=False)
+            webpage_url = self._search_regex(
+                r'<a class="tgme_widget_message_video_player[^>]+href="([^"]+)"',
+                video, 'webpage URL', fatal=False)
+            if not video_url or not webpage_url:
+                continue
+            formats = [{
+                'url': video_url,
+                'ext': 'mp4',
+            }]
+            self._sort_formats(formats)
+            videos.append({
+                'id': url_basename(webpage_url),
```

Comment:
```suggestion
                'id': url_basename(webpage_url),
                'webpage_url': update_url_query(webpage_url, {'single': True}),
```

Playlist entries should have `webpage_url` if possible. When returning `url_result`, core code handles this, but when directly returning video, extractor has to manually add it